crates/provider/src/native.rs

```diff
@@ -2,531 +2,920 @@ use std::{
-}
-
-pub struct Node {}
-
-#[async_trait]
-impl<T> Provider for NativeProvider<T>
+pub struct NativeProviderOptions<FS>
 where
-    T: FileSystem + Send + Sync,
+    FS: FileSystem + Send + Sync,
 {
-    type Node = Node;
+    filesystem: FS,
+    tmp_dir: Option<String>,
+}
 
-    fn require_image() -> bool {
-        false
-    }
+#[derive(Debug)]
+struct NativeProviderInner<FS: FileSystem + Send + Sync + Clone> {
+    capabilities: ProviderCapabilities,
+    tmp_dir: String,
+    namespaces: HashMap<String, NativeNamespace<FS>>,
+    filesystem: FS,
+}
 
-    async fn create_namespace(&mut self) -> Result<(), ProviderError> {
-        // Native provider don't have the `namespace` isolation.
-        // but we create the `remoteDir` to place files
-        self.filesystem
-            .create_dir(&self.remote_dir)
-            .await
-            .map_err(|e| ProviderError::FSError(Box::new(e)))?;
-        Ok(())
-    }
+#[derive(Debug, Clone)]
+pub struct NativeProvider<FS: FileSystem + Send + Sync + Clone> {
+    inner: Arc<RwLock<NativeProviderInner<FS>>>,
+}
 
-    async fn destroy_namespace(&self) -> Result<(), ProviderError> {
-        // get pids to kill all related process
-        let pids: Vec<String> = self
-            .process_map
-            .iter()
-            .filter(|(_, process)| process.pid != 0)
-            .map(|(_, process)| process.pid.to_string())
-            .collect();
-
-        // TODO: use a crate (or even std) to get this info instead of relying on bash
-        let result = self
-            .run_command(
-                [format!(
-                    "ps ax| awk '{{print $1}}'| grep -E '{}'",
-                    pids.join("|")
-                )]
-                .to_vec(),
-                NativeRunCommandOptions {
-                    is_failure_allowed: true,
+#[derive(Debug, Clone)]
+struct WeakNativeProvider<FS: FileSystem + Send + Sync + Clone> {
+    inner: Weak<RwLock<NativeProviderInner<FS>>>,
+}
+
+impl<FS: FileSystem + Send + Sync + Clone> NativeProvider<FS> {
+    pub fn new(options: NativeProviderOptions<FS>) -> Self {
+        NativeProvider {
+            inner: Arc::new(RwLock::new(NativeProviderInner {
+                capabilities: ProviderCapabilities {
+                    requires_image: false,
                 },
-            )
-            .await
-            .unwrap();
-
-        if result.exit_code.code().unwrap() == 0 {
-            let pids_to_kill: Vec<String> = result
-                .std_out
-                .split(|c| c == '\n')
-                .map(|s| s.into())
-                .collect();
-
-            let _ = self
-                .run_command(
-                    [format!("kill -9 {}", pids_to_kill.join(" "))].to_vec(),
-                    NativeRunCommandOptions {
-                        is_failure_allowed: true,
-                    },
-                )
-                .await?;
+                tmp_dir: options.tmp_dir.unwrap_or(DEFAULT_TMP_DIR.to_string()),
```

Comment:
We should get the `tmp_dir` from the OS if is not provided.

Reply:
What do you mean, is there a way to detect the "default tmp dir" for an os ?